V2: Regl renderer typescript definitions and generator.

kenianbei opened the pull request:
### All Submissions:

* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [x] Does your submission pass tests?


> Note: please do NOT include build files (those generate by the build-xxx commands) with your PR,

Thank you for your help in advance, much appreciated !

kenianbei commented:
I added a types generator command to npm scripts, it works fairly well and I only modified a couple lines to get some of the regl definitions. Doesn't use js docs since regl-renderer looks to be still in development.

z3dev commented:
@kenianbei quick question... why opt for the generation of definitions? there won't be many change to the definitions.

kenianbei commented:
Just for convenience in case the code changed. Is it preferable to remove it from scripts and not included the tsconfig?

Also, after trying it out in my project there are still a bunch of "any" types that would be nice to have typed, I think I'll go through and type all the ones the generator missed when I have time this week.

z3dev commented:
> Just for convenience in case the code changed. Is it preferable to remove it from scripts and not included the tsconfig?

From my understanding (based on the modeling package), the generation of definitions will be kind of lame. If there’s manual changes then probably the generation should be removed to prevent overwriting changes. If the generation is spot on then let’s keep it.

Maybe @danmarshall can add a few words of wisdom here.

z3dev commented:
> Just for convenience in case the code changed. Is it preferable to remove it from scripts and not included the tsconfig?
> 
> Also, after trying it out in my project there are still a bunch of "any" types that would be nice to have typed, I think I'll go through and type all the ones the generator missed when I have time this week.

@kenianbei can you complete the changes? (See above) Also, your branch needs to up to date with the latest from the master branch.

kenianbei commented:
I removed the auto type generation and fixed some types. We will definitely want to revisit this if it is cleaned up and has proper js docs. But for now it probably is good enough for most use cases.

z3dev commented on packages/utils/regl-renderer/types/rendering/commands/drawMesh/colorOnlyShaders.d.ts:
```diff
@@ -0,0 +1,2 @@
+declare const vColorFrag: "\nprecision mediump float;\nuniform vec4 ucolor;\n\nvoid main () {\n  gl_FragColor = ucolor;\n}\n";
```
UG… should these be exposed? These shader declarations seem to be internal only.

kenianbei replied:
It's exported so it's public. When I auto generated the types it only generates types for exported members.

kenianbei replied:
I've removed this since it's not used in another type.

z3dev commented on packages/utils/regl-renderer/types/bound-utils/computeBounds.test.d.ts:
```diff
@@ -0,0 +1 @@
+export {};
```
Why a declaration for tests? Can this be removed?

kenianbei replied:
Same as above...

z3dev commented on packages/utils/regl-renderer/types/geometry-utils-V1/cagToGeometries.d.ts:
```diff
@@ -0,0 +1,8 @@
+export = cagToGeometries;
```
Please remove as V1 is not supported / tested. These functions will be removed in the future.

kenianbei replied:
Removed.

z3dev commented on packages/utils/regl-renderer/types/index.d.ts:
```diff
@@ -0,0 +1,18 @@
+export namespace controls {
+    const orbit: typeof import("./controls/orbitControls");
+}
+export const entitiesFromSolids: (params: { color?: number[]; smoothNormals?: boolean }, solids: Geom2[] | Geom3[]) => Geometry[];
```
Can the declaration of entitiesFromSolids() be imported?

kenianbei replied:
Yes, I've tested it in jscad-react to make sure.

z3dev reviewed (changes requested):
Sorry for the late comments. Please review the comments and adjust the TS declarations.

z3dev commented on packages/utils/regl-renderer/types/geometry-utils-V2/entitiesFromSolids.test.d.ts:
```diff
@@ -0,0 +1 @@
+export {};
```
Drop?

z3dev commented on packages/utils/regl-renderer/types/geometry-utils-V2/geom2ToGeometries.test.d.ts:
```diff
@@ -0,0 +1 @@
+export {};
```
Drop?

z3dev commented on packages/utils/regl-renderer/types/geometry-utils-V2/geom3ToGeometries.test.d.ts:
```diff
@@ -0,0 +1 @@
+export {};
```
Drop?

z3dev commented on packages/utils/regl-renderer/types/geometry-utils-V2/path2ToGeometries.test.d.ts:
```diff
@@ -0,0 +1 @@
+export {};
```
Drop?

z3dev commented on packages/utils/regl-renderer/types/rendering/commands/drawMesh/meshShaders.d.ts:
```diff
@@ -0,0 +1,3 @@
+declare const meshVert: "\nprecision mediump float;\n\nuniform float camNear, camFar;\nuniform mat4 model, view, projection;\n\nattribute vec3 position, normal;\n\nvarying vec3 surfaceNormal, surfacePosition;\nvarying vec4 _worldSpacePosition;\n\nvoid main() {\n  surfacePosition = position;\n  surfaceNormal = normal;\n  vec4 worldSpacePosition = model * vec4(position, 1);\n  _worldSpacePosition = ...
```
The redeclaration of these massive string seems wasteful. Are these declarations useful?

z3dev commented on packages/utils/regl-renderer/types/rendering/commands/drawMesh/vColorShaders.d.ts:
```diff
@@ -0,0 +1,3 @@
+declare const vColorFrag: "\nprecision mediump float;\nvarying vec3 surfaceNormal, surfacePosition;\n\nuniform float ambientLightAmount;\nuniform float diffuseLightAmount;\nuniform float specularLightAmount;\n\nuniform vec3 lightDirection;\nuniform vec4 lightColor;\nuniform vec3 opacity;\nuniform float uMaterialShininess;\n\nvarying vec4 vColor;\nuniform vec4 ucolor;\nuniform float vColorToggler; ...
```
The redeclaration of these massive string seems wasteful. Are these declarations useful?

z3dev reviewed (changes requested):
So close! Sorry to be picky but there are a few more changes.

z3dev commented:
@kenianbei can you spend some time to make the final changes? these changes would be great to have in the next release.

kenianbei commented:
Will try to get time this week to finish. May just remove everything but the main render API for simplicity of maintenance.

z3dev commented:
> Will try to get time this week to finish. May just remove everything but the main render API for simplicity of maintenance.

Super. And thanks beforehand.

FYI, there have been changes… lots… so please sync up with the master branch. 😃

kenianbei commented:
Hey, I think it should be ready now, I tested it with jscad-react. Let me know if there is any other changes to do, I should have time this week if there is.

z3dev commented:
@hrgdavor can you review these changes for the TS definitions? I believe that you use TS sometimes.

z3dev commented on packages/utils/regl-renderer/types/controls/orbitControls.d.ts:
```diff
@@ -0,0 +1,125 @@
+import { Geometry } from '@jscad/modeling/src/geometries/types'
```
Is this import necessary? I would be surprised if so.

kenianbei replied:
Not necessary, but it's nice... for example when you use entitiesFromSolids() it will let you know that the return  of that function is type Geometry[], which will show up in an IDE with correct typed properties like polygons and transforms. If you type it as any then you don't get autocomplete on properties or type checking for other functions like zoomToFit.

z3dev replied:
Ok

z3dev reviewed (commented):
@kenianbei SUPER! This looks really nice.

I found one small issue. Let me know what you think.

hrgdavor commented:
> @hrgdavor can you review these changes for the TS definitions? I believe that you use TS sometimes.

@z3dev unfortunately I do not have experience with TS. I am oldschool JS dude 😄 
I was curious about a good setup that would wotk with old style js and TS for some transition, and somebody in one of the issues here promissed to share a good TS setup ... but that never happened.

Sorry, I am of no use here.

z3dev reviewed (approved):
Super! Thanks for these definitions.

Please feel free to contribute further. :)
